File: src/main/java/com/tang/designpatterns/chap1/java/Main.java
package com.tang.designpatterns.chap1.java;


/**
 * @Title: Main
 * @Description:
 * @author: tangyao
 * @date: 2022/9/28 10:36
 * @Version: 1.0
 */

public class Main {

    /**
     * 迭代器模式： 引入 Iterator 后 可将遍历与实现分离开
     * * 这里的while循环 并不依赖于BookSelf的实现
     *
     * @param args
     */
    public static void main(String[] args) {
        BookShelf bookShelf = new BookShelf(4);
        for (int i = 0; i < 10; i++) {
            bookShelf.appendBook(new Book("Around the World int 80 Days"));
            bookShelf.appendBook(new Book("Bible"));
            bookShelf.appendBook(new Book("Cinderella"));
            bookShelf.appendBook(new Book("Daddy-Long-Legs"));
        }

        Iterator it = bookShelf.iterator();
        while (it.hasNext()) {
            Book book = (Book) it.next();
            System.out.println(book.getName());
        }
    }
}

File: src/main/java/com/tang/designpatterns/chap1/scala/Main.scala
package com.tang.designpatterns.chap1.scala


@main def main =
  val bookShelf = new BookShelf(4)
    bookShelf.appendBook(new Book("Around the World int 80 Days"))
    bookShelf.appendBook(new Book("Bible"))
    bookShelf.appendBook(new Book("Cinderella"))
    bookShelf.appendBook(new Book("Daddy-Long-Legs"))

//  (0 until 100).foreach(i =`> {
//    bookShelf.appendBook(new Book(String.valueOf(i)))
//  });

  val it = bookShelf.iterator()
  while (it.hasNext()) {
    val book: Book = it.next().asInstanceOf[Book]
    println(book.name)
  }

  val it2 = bookShelf.reverseIterator()
  while it2.hasNext() do
    val book = it2.next().asInstanceOf[Book]
    if book == null then
      println()

    println (book.name)

